<a href="https://colab.research.google.com/github/VindelM/Endometriosis/blob/main/Endometriosis_type_shii.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Problem Statement
This project aims to develop an advance machine learning system for predicting the onset of endometriostis in the reproductive age of women. The inability to detect illness with this traditional methods such laparoscopic and other techniques are time consuming, invasive and cost intensive. Hence, there a need to leverage AI models to detect this illness with low cost  WITH QUIC RESPONSE TIME.

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
import warnings
warnings.filterwarnings('ignore')
from sklearn.feature_selection import SelectKBest, chi2
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier, BaggingClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import SGDClassifier
from xgboost import XGBClassifier

In [ ]:
df = pd.read_csv('structured_endometriosis_data.csv')
df

Age  Menstrual_Irregularity  Chronic_Pain_Level  \
0      24                       1            8.361531   
1      37                       1            4.995508   
2      46                       1            3.363996   
3      32                       1            5.246037   
4      28                       1            3.898932   
...   ...                     ...                 ...   
9995   37                       1            3.153169   
9996   44                       0            4.044800   
9997   39                       0            5.096384   
9998   47                       1            7.598862   
9999   38                       0            7.822210   

      Hormone_Level_Abnormality  Infertility        BMI  Diagnosis  
0                             0            0  19.451314          0  
1                             0            0  22.388436          0  
2                             1            0  21.320443          0  
3                             0            0  20.177715          1  
4                             1            0  23.538103          1  
...                         ...          ...        ...        ...  
9995                          1            0  18.318849          0  
9996                          1            1  24.732344          0  
9997                          1            1  34.204883          1  
9998                          1            1  30.374964          1  
9999                          0            0  26.385575          0  

[10000 rows x 7 columns]

In [ ]:
df.duplicated().sum()

np.int64(0)

# Task
Train a logistic regression model to achieve 95% accuracy.

## Prepare data

### Subtask:
Separate the features and the target variable, and split the data into training and testing sets.


**Reasoning**:
Separate the features and target variable, then split the data into training and testing sets.



In [ ]:
X = df.drop('Diagnosis', axis=1)
y = df['Diagnosis']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Scale data

### Subtask:
Scale the features using StandardScaler.


**Reasoning**:
Scale the training and testing features using StandardScaler as instructed.



In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

## Train logistic regression model

### Subtask:
Train a Logistic Regression model on the training data.


**Reasoning**:
Instantiate and train a Logistic Regression model on the scaled training data.



In [ ]:
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

## Evaluate model

### Subtask:
Evaluate the model's performance on the test data and check for 95% accuracy. If the accuracy is not 95%, iterate on hyperparameter tuning or try different models.


**Reasoning**:
Evaluate the performance of the trained logistic regression model on the test data by making predictions and calculating the accuracy.



In [ ]:
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Model Accuracy: {accuracy:.4f}")

Model Accuracy: 0.6355


**Reasoning**:
The current model accuracy is 0.6355, which is below the target of 0.95. According to the instructions, I need to try other models to improve the accuracy. I will try training a RandomForestClassifier model.



In [ ]:
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f"Random Forest Model Accuracy: {accuracy_rf:.4f}")

Random Forest Model Accuracy: 0.5940


**Reasoning**:
The Random Forest model did not achieve the target accuracy. I will try another model, the Gradient Boosting Classifier, to see if it performs better.



In [ ]:
gb_model = GradientBoostingClassifier(random_state=42)
gb_model.fit(X_train, y_train)
y_pred_gb = gb_model.predict(X_test)
accuracy_gb = accuracy_score(y_test, y_pred_gb)
print(f"Gradient Boosting Model Accuracy: {accuracy_gb:.4f}")

Gradient Boosting Model Accuracy: 0.6330


**Reasoning**:
None of the models trained so far (Logistic Regression, Random Forest, Gradient Boosting) have reached the target accuracy of 95%. According to the instructions, if the accuracy is not met, I should continue trying different models from the imported list. I will try a Decision Tree Classifier next.



In [ ]:
dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)
y_pred_dt = dt_model.predict(X_test)
accuracy_dt = accuracy_score(y_test, y_pred_dt)
print(f"Decision Tree Model Accuracy: {accuracy_dt:.4f}")

Decision Tree Model Accuracy: 0.5585


**Reasoning**:
The Decision Tree Classifier also did not achieve the target accuracy. I will try the Bagging Classifier next, as it can sometimes improve performance by combining multiple base models.



In [ ]:
bagging_model = BaggingClassifier(random_state=42)
bagging_model.fit(X_train, y_train)
y_pred_bagging = bagging_model.predict(X_test)
accuracy_bagging = accuracy_score(y_test, y_pred_bagging)
print(f"Bagging Model Accuracy: {accuracy_bagging:.4f}")

Bagging Model Accuracy: 0.5945


**Reasoning**:
The Bagging Classifier did not achieve the target accuracy. I will try the XGBoost Classifier next, which is known for its strong performance in many classification tasks.



In [ ]:
xgb_model = XGBClassifier(random_state=42)
xgb_model.fit(X_train, y_train)
y_pred_xgb = xgb_model.predict(X_test)
accuracy_xgb = accuracy_score(y_test, y_pred_xgb)
print(f"XGBoost Model Accuracy: {accuracy_xgb:.4f}")

XGBoost Model Accuracy: 0.6030


## Evaluate model

### Subtask:
Evaluate the model's performance on the test data and check for 95% accuracy. If the accuracy is not 95%, iterate on hyperparameter tuning or try different models.


## Summary:

### Data Analysis Key Findings
*   The initial Logistic Regression model achieved an accuracy of 63.55% on the test data.
*   Several other models were attempted, including Random Forest (59.40% accuracy), Gradient Boosting (63.30% accuracy), Decision Tree (55.85% accuracy), Bagging (59.45% accuracy), and XGBoost (60.30% accuracy).
*   None of the models trained achieved the target accuracy of 95%.

### Insights or Next Steps
*   Further feature engineering or data collection might be necessary to improve model performance.
*   Exploring more complex modeling techniques or ensembling methods could potentially yield higher accuracy, although reaching 95% might be challenging with this dataset.
